Add custom event on h5peditor when it is ready to use

SergeyYukhnevich opened the pull request:
This fix will add ability to listen when h5p editor is ready and use it api.
e.g. listen for library change etc.

you can listen this event outside of h5p editor: `window.addEventListener('h5peditor-ready', callback);`

falcon-git commented:
I have an idea of how this could be useful, but still, exactly what and how are you using this?

SergeyYukhnevich commented:
hi @falcon-git ,
I have functionality that should work for specific libraries e.g. (question set).
When this library is selected I should enable this functionality.
So I've triggered event when h5p iframe is ready to use and pass editor object to get access to the hub select.

falcon-git commented:
Ok, which platform are you on?

SergeyYukhnevich commented:
Drupal

H5P-Helper commented:
Have you looked into this?

https://h5p.org/drupal-customization

Kind regards,

Falcon

for Joubel

h5p/h5p-editor-php-library on May 30, 2017 at 1:31pm wrote:

Drupal

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub 
( https://github.com/h5p/h5p-editor-php-library/pull/36#issuecomment-304850730 )
, or mute the thread 
( https://github.com/notifications/unsubscribe-auth/ATkhbHSVODnCVuPSknSKaUmIArw11bE4ks5r-_4NgaJpZM4Npchg )
.

h5p/h5p-editor-php-library on May 30, 2017 at 1:30pm wrote:

Ok, which platform are you on?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub 
( https://github.com/h5p/h5p-editor-php-library/pull/36#issuecomment-304850586 )
, or mute the thread 
( https://github.com/notifications/unsubscribe-auth/ATkhbOGbRGlkQcqo3TprurJ0Kno6cSSKks5r-_3dgaJpZM4Npchg )
.

h5p/h5p-editor-php-library on May 30, 2017 at 1:29pm wrote:

hi @falcon-git ( https://github.com/falcon-git ) ,

I have functionality that should work for specific libraries e.g.
(question set).

When this library is selected I should enable this functionality.

So I've triggered event when h5p iframe is ready to use and pass
editor object to get access to the hub select.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub 
( https://github.com/h5p/h5p-editor-php-library/pull/36#issuecomment-304850412 )
, or mute the thread 
( https://github.com/notifications/unsubscribe-auth/ATkhbPpjmXf2_TlY9Az8tfqKPNU16oQsks5r-_2sgaJpZM4Npchg )
.

SergeyYukhnevich commented:
@H5P-Helper thank you for your reply, I've tried to use the api that you gave me, but unfortunately it doesn't fit my needs

falcon-git commented:
Ok, it at least allow you to add a script into the frame so that you won't have to have your script in the parent.

SergeyYukhnevich commented:
Yes, I can add script but it does not help me because I need to listen for library change anyway.

If i use that hook: `hook_h5p_scripts_alter` I can't get library by machine name in edit mode e.g. `isset($libraries['H5P.QuestionSet'])`, this won't work.

If I use that hook: `hook_h5p_semantics_alter` I have `machine_name` but still it doesn't help me :(

I think that there should be some way from js to interact with library select externally and listen for some events like `h5peditor-ready` or `library-selected`, don't you think so?

falcon-git commented:
Ok, so you need to do something in the parent window when the library changes in the H5P Editor iFrame window?

Yes, I agree, there definitively should be a nice API for this in js. For some reason I assumed that you wanted to make changes inside the H5P iframe and then scripts_alter would be useful since it would allow you to easily run your script inside the iframe, but having the event triggered on the parent is probably at least as useful. I think we should create a system that does both, and that can easily be extended to include more events.

falcon-git commented:
It probably should work more or less the same way as the external dispatcher with events bubbling up and being available throughout the tree ending up in the parent window.

SergeyYukhnevich commented:
@falcon-git Yes, I agree.
If you want we can collaborate on that issue, because this external event is very useful.
Maybe you can suggest cleaner solution or the one that use external dispatcher?

SergeyYukhnevich commented:
Good morning @falcon-git,

I've changed a bit my code, now I'm triggering the event on Iframe parent element with that id: `#edit-h5p-editor`.

Also I've changed ajax request structure from chained functions to object literal syntax, because it is more readable.

icc commented:
Hi Sergey, 
This looks like a good solution, but I'm not 100% sure of the use-case here. Would you be able to provide an example of how you're planning to use the new event? Just to ensure that we're getting the best and most flexible solution for the problem.

SergeyYukhnevich commented:
Hello @icc,
Here is an example of usage.

```js
document.getElementById('edit-h5p-editor').addEventListener('h5peditor-ready', function(event) {
  // get h5p selector hub instance.
  var select = event.detail.selector.selector;

  // check if selected library has right type to enable external feature.
  checkType(select.currentLibrary, context);

  // Listen to hub event to detect library change.
  select.on('selected', function() {
    checkType(select.currentLibrary, context);
  });

  // Listen to client hub event to detect library change.
  select.client.on('clear-upload-form', function() {
    checkType(select.currentLibrary, context);
  });
});
```

icc commented:
Thanks for sharing @SergeyYukhnevich 
What worries me is what the external feature might be – is it something that will affect or make changes to the H5P content in any way? Or is it just like adding or adapting some of the other node form fields?

One more consideration is that in the upcoming modules/plugins there will be multiple editors in the same form. So the editor element must be targeted in a scalable way.

SergeyYukhnevich commented:
@icc , the external feature doesn't modify content inside iframe.
It just enables certificate option in our system if one of supported libraries loaded e.g. H5P.QuestionSet.

icc commented:
I've done some thinking on the matter and there's mainly two concerns:

1) We don't wish to expose more API than we have too, this will just be crippeling when we start to refactor editor parts.
Since all H5P events are currently exposed through the `H5P.externalDispatcher` property, do you think using that would be a feasible option for you? E.g
```
H5P.externalDispatcher.on('editorLoaded', function (event) {
  var library = event.data.split(' ')[0]; 

  if (event.data === 'H5P.QuestionSet') {
    // Do stuff
  }
})
```

2) When do we need the event? E.g. should we get it right away as the library is selected or should we wait until the editor is actually loaded(around line _167_ in `h5peditor-library-selector.js`)… or we could do both. (the loading of the content type could fail)
If it's before I think we need to add the triggering before all places in `h5peditor-library-selector.js`(inside iframe) where `that.selector.getSelectedLibrary(librarySelectHandler);` gets called. 

The relaying of the event needs to happen inside the AJAX in `h5peditor-editor.js`(outside iframe), where you currently have the dispatchEvent. E.g.
```
self.selector.on('editorLoaded', function (event) {
  H5P.externalDispatcher.trigger('editorLoaded', event);
});
```

Also, we don't do `'use strict';` on the rest of the editor and we would like it to stay consistent across all the code.

SergeyYukhnevich commented:
@icc I've used `H5P.externalDispatcher` and triggered several events on editor load and on library select, please check
